### PR TITLE
Refactor/Optimize Linear Regression Modeling Task

### PR DESCRIPTION
**Pull request by inf0rmatiker**
## Notes

Big PR here. Below is a list of items I've implemented today:

- [Builder Design Pattern](https://www.stackabuse.com/the-builder-design-pattern-in-java/) applied Linear Regression model construction. Please read that link to see why it applies to our situation. Should be extended to all other models, and crude validation should be performed at build time, matching the parameters defined in our `model_catalogue` collection.
- Intermediate abstract class `ModelHandler`, extending `GrpcHandler<ModelRequest, ModelResponse>`. All models should extend/implement this. Specifically, the `isValid()` function for evaluating the `ModelRequest`'s validity.
- `JavaSparkContext` pulled all the way out to the `modelQuery()` entrypoint. Instance shared between model types. @MillerRidgeway  added specific TODO comments hopefully pointing you in the right direction.
- Dependency JARs are now added at `modelQuery()` time, instead of within your model classes (that kind of code doesn't belong there).
- Reworked what is actually needed to build a model. For regression models, a Mongo collection (`Dataset<Row>`) can be input and shared across _all models_ built for the individual GISJoins. No need to tell the models at all about the Mongo or Spark connector instances.

## TODO

This PR is still in progress. Things to add still:

- Move some of Menuka/Saptashwa's models over to new design structure.
- Add a profiler util for taking time metrics on our model building operations, among other things.

### CODE REVIEW

File: src/main/java/org/sustain/server/SustainServer.java [hunk 1]
```diff
@@ -162,36 +165,47 @@ public void modelQuery(JsonModelRequest request,
 
         @Override
         public void modelQuery(ModelRequest request, StreamObserver<ModelResponse> responseObserver) {
-            GrpcHandler<ModelRequest, ModelResponse> handler;
+            SparkSession sparkSession = SparkSession.builder()
+                    .master(Constants.Spark.MASTER)
+                    .appName("SUSTAIN Linear Regression Model")
```
Comment (inf0rmatiker) [inline src/main/java/org/sustain/server/SustainServer.java:170]: whoopsie

File: src/main/java/org/sustain/handlers/RegressionQueryHandler.java [hunk 2]
```diff
@@ -1,33 +1,83 @@
...
             logRequest(this.request);
-            LinearRegressionRequest req = this.request.getLinearRegressionRequest();
-            for (String gisJoin: req.getGisJoinsList()) {
-                LinearRegressionResponse modelResults = buildModel(this.request, gisJoin);
+
+            // Set parameters of Linear Regression Model
+            LinearRegressionRequest lrRequest = this.request.getLinearRegressionRequest();
+            Collection requestCollection = this.request.getCollections(0); // We only support 1 collection currently
+
+            // Create a custom ReadConfig
```
Comment (inf0rmatiker) [inline src/main/java/org/sustain/handlers/RegressionQueryHandler.java:41]: This is how Spork suggests doin it: https://docs.mongodb.com/spark-connector/master/java/read-from-mongodb/

File: src/main/java/org/sustain/modeling/LinearRegressionModelImpl.java [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+                .forLabel("max_max_air_temperature")
+                .withMaxIterations(100)
+                .withEpsilon(1.35)
+                .withTolerance(1E-7)
+                .build();
+
+
+        lrModel.buildAndRunModel();
+        log.info("Executed LinearRegressionModelImpl.main() successfully");
+        sparkContext.close();
```
Comment (MillerRidgeway) [inline src/main/java/org/sustain/modeling/LinearRegressionModelImpl.java:193]: We'll need to make sure to come back and update these experiments - if anyone ever ends up running them they _could_ kill all the running spark jobs on our dev cluster unintentionally (similar to the problem we have now).

File: src/main/java/org/sustain/server/SustainServer.java [hunk 4]
```diff
@@ -162,36 +165,47 @@ public void modelQuery(JsonModelRequest request,
...
+                    .config("spark.mongodb.input.uri", String.format("mongodb://%s:%d", Constants.DB.HOST, Constants.DB.PORT))
+                    .config("spark.mongodb.input.database", Constants.DB.NAME)
+                    .getOrCreate();
+
+            // TODO: MILLER: this should somehow be pull out and exist across invocations of modelQuery()
```
Comment (MillerRidgeway) [inline src/main/java/org/sustain/server/SustainServer.java:175]: I think @hamersaw was planning on attacking this, but I see no reason why the context can't be pulled out into a static constant like `Constants.sparkContext` if you can easily perform an on-the-fly override like in [`RegressionQueryHandler`](https://github.com/Project-Sustain/sustain-query-service/blob/0f75d22fedfe206993ea58bcfee9d3f72f7d32e8/src/main/java/org/sustain/handlers/RegressionQueryHandler.java#L42-L44)

File: src/main/java/org/sustain/modeling/LinearRegressionModelImpl.java (hunk 3 above)
Comment (inf0rmatiker) [reply src/main/java/org/sustain/modeling/LinearRegressionModelImpl.java:193]: I'm not sure if running this from a separate machine (or even just a separate process) would conflict with _all_ other spark contexts. I'll have to try this and see if it breaks it.

**Review by hamersaw — COMMENTED** [review]
How soon can we get this merged? The overall pull request is a little ridiculous, refactoring and optimizing the linear regression task shouldn't need to touch 26 files.

**Comment by inf0rmatiker** [comment]
> How soon can we get this merged? The overall pull request is a little ridiculous, refactoring and optimizing the linear regression task shouldn't need to touch 26 files.

I agree I went a little off the rails with this PR, ended up implementing that static spark context myself since there seemed to be some confusion about who was gonna tackle that.

I'll test it this morning and let you know the results.

**Comment by hamersaw** [comment]
> > How soon can we get this merged? The overall pull request is a little ridiculous, refactoring and optimizing the linear regression task shouldn't need to touch 26 files.
> 
> I agree I went a little off the rails with this PR, ended up implementing that static spark context myself since there seemed to be some confusion about who was gonna tackle that.
> 
> I'll test it this morning and let you know the results.

Sounds good. After looking into the whole SparkContext ordeal, it may be a bit complicated. I'm still thinking about it. Submitting concurrent Spark jobs is easy, doing so correctly may require a more elegant solution. For example, by sharing the SparkContext between all jobs, an error in one job will kill all other running jobs.

**Pull request merged**